moshi/src/test/java/com/squareup/moshi/EnumMapJsonAdapterTest.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (C) 2015 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.moshi;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.EnumMap;
+import okio.Buffer;
+import org.assertj.core.data.MapEntry;
+import org.junit.Test;
+
+import static com.squareup.moshi.TestUtil.newReader;
+import static com.squareup.moshi.Util.NO_ANNOTATIONS;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.fail;
+
+public final class EnumMapJsonAdapterTest {
+  private final Moshi moshi = new Moshi.Builder().build();
+
+  @Test public void map() throws Exception {
+    EnumMap<Rating, String> map = new EnumMap<>(Rating.class);
+    map.put(Rating.G, "Lion King");
+    map.put(Rating.PG_13, "Scream");
+    map.put(Rating.R, null);
+    map.put(Rating.NC_17, "Saw");
+
+    String toJson = toJson(Rating.class, String.class, map);
+    assertThat(toJson).isEqualTo("{"
+        + "\"G\":\"Lion King\","
+        + "\"PG_13\":\"Scream\","
+        + "\"R\":null,"
+        + "\"NC_17\":\"Saw\""
+        + "}");
+
+    EnumMap<Rating, String> fromJson = fromJson(Rating.class, String.class, toJson);
+    assertThat(fromJson).containsExactly(
+        MapEntry.entry(Rating.G, "Lion King"), MapEntry.entry(Rating.PG_13, "Scream"),
+        MapEntry.entry(Rating.R, null), MapEntry.entry(Rating.NC_17, "Saw"));
+  }
+
+  @Test public void emptyMap() throws Exception {
+    EnumMap<Rating, Boolean> map = new EnumMap<>(Rating.class);
+
+    String toJson = toJson(Rating.class, String.class, map);
+    assertThat(toJson).isEqualTo("{}");
+
+    EnumMap<Rating, Boolean> fromJson = fromJson(Rating.class, String.class, "{}");
+    assertThat(fromJson).isEmpty();
+  }
+
+  @Test public void nullMap() throws Exception {
+    JsonAdapter<?> jsonAdapter = enumMapAdapter(Rating.class, String.class);
+
+    Buffer buffer = new Buffer();
+    JsonWriter jsonWriter = JsonWriter.of(buffer);
+    jsonWriter.setLenient(true);
+
+    jsonAdapter.toJson(jsonWriter, null);
+    assertThat(buffer.readUtf8()).isEqualTo("null");
+
+    JsonReader jsonReader = newReader("null");
+    jsonReader.setLenient(true);
+    assertThat(jsonAdapter.fromJson(jsonReader)).isEqualTo(null);
+  }
+
+  @Test public void orderIsRetained() throws Exception {
```

Comment:
This test is insincere. Order is only maintained because you are using the declaration order.